docs/proposals/ingress-l4.md

```diff
@@ -0,0 +1,346 @@
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Ingress L4 Proposal
+
+## Abstract
+
+A proposal to add support for TCP and UDP to the Ingress Resource.
+
+Existing issue regarding Ingress and TCP
+* Ingress and TCP [#23291](https://github.com/kubernetes/kubernetes/issues/23291)
+
+
+## Use Cases
+
+1. Be able to accept and route traffic for arbitrary TCP ports
+1. Be able to accept and route traffic for arbitrary UDP ports
+
+
+## Motivation
+
+When deploying and exposing non-http(s) based services, Kubernetes users are
+limited to NodePort and service type=LoadBalancer as methods for exposing these
+services using interfaces supported by the Kubernetes API.
+
+Other methods exist, via contributed code, for exposing non-http(s) based services:
+
+* [keepalived-vip](https://github.com/kubernetes/contrib/tree/master/keepalived-vip)
+* NGINX Ingress Controller
+  [nginx-ingress](https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx)
+
+However these implementations rely on a ConfigMap being populated with a
+mapping of destination endpoint to service.  A single common interface for
+describing all Ingress would is highly desirable.
+
+
+## Design Considerations
+
+### State of L4 Protocol Support in Cloud Providers
+
+L4 support is not uniform across all Cloud Providers, specifically TLS
+termination is lacking in several Cloud Providers.
+
+|           | TCP | UDP | TLS |
+| --------- | --- | --- | --- |
+| AWS       |  x  |     |  x  |
+| GCE       |  x  |  x  | (1) |
+| Mesos     |  ?  |  ?  |  ?  |
+| Openstack |  x  |  x  |  x  |
+| Rackspace |  x  |  x  |     |
+| vSphere   |  x  |  x  |  x  |
+
+(1) GCE has TLS as an alpha service
+
+### TCP and UDP support
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+    - udp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+```
+
+While HTTP rules have host specifiers, in TCP and UDP rules they would not be
+allowed as hostname handling is specific to L5 or higher layer protocol being
+used over TCP or UDP.  TCP and UDP as protocol specifiers should be agnostic to
+higher layer protocols.
+
+### Ingress type specifier
+
+The current Ingress API assumes a protocol based on the presence or absence
+of a key (tls, http).  This leaves open the possibility of a spec that defines
+all protocols.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
```

Comment:
I think this would be fine if rules was viewed as a list of maps where the key is the protocol type like you showed.  The thing I found confusing about the current syntax is how tls relates to the ingress rule as specified in http://kubernetes.io/docs/user-guide/ingress/

For example
`
rules:
- tcp:
    # defaults to backend port, errs on collision or if both unspecified
    port: 3306
    backend:
      serviceName: testmysql
      # defaults to frontend port
      port: 3306
- tls:
    cert:
    ...     
- http:
    paths:
    - path: /bar
      backend: 
        servcieName: bar
- https:
    cert:
    ...
    paths:
    ...
...
`

or something similar to this

My reason for prefering a protocol specifier is that a grammar like the one above potentially allows something like this

`
- tcp:
    # defaults to backend port, errs on collision or if both unspecified
    port: 3306
    backend:
      serviceName: testmysql
      # defaults to frontend port
      port: 3306
  tls:
    cert:
    ...     
  http:
    paths:
    - path: /bar
      backend: 
        servcieName: bar
`

while this can be disallowed programmatically, the syntax/grammar doesn't make it obvious.